precompiles/randomness/src/lib.rs

```diff
@@ -0,0 +1,531 @@
+	RelayEpochIndex = "relayEpochIndex()",
+	RequestBabeRandomnessCurrentBlock =
+		"requestBabeRandomnessCurrentBlock(address,uint256,uint64,bytes32,uint64)",
+	RequestBabeRandomnessOneEpochAgo =
+		"requestBabeRandomnessOneEpochAgo(address,uint256,uint64,bytes32)",
+	RequestBabeRandomnessTwoEpochsAgo =
+		"requestBabeRandomnessTwoEpochsAgo(address,uint256,uint64,bytes32)",
+	RequestLocalRandomness = "requestLocalRandomness(address,uint256,uint64,bytes32,uint64)",
+	FulfillRequest = "fulfillRequest(uint64)",
+	IncreaseRequestFee = "increaseRequestFee(uint64)",
+	ExecuteRequestExpiration = "executeRequestExpiration(uint64)",
+	InstantBabeRandomnessCurrentBlock = "instantBabeRandomnessCurrentBlock(uint64,bytes32)",
+	InstantBabeRandomnessOneEpochAgo = "instantBabeRandomnessOneEpochAgo(uint64,bytes32)",
+	InstantBabeRandomnessTwoEpochsAgo = "instantBabeRandomnessTwoEpochsAgo(uint64,bytes32)",
+	InstantLocalRandomness = "instantLocalRandomness(uint64,bytes32)",
+}
+
+pub const FULFILLMENT_ESTIMATED_COST: u64 = 1000u64; // TODO: get real value from benchmarking
+pub const LOG_SUBCALL_SUCCEEDED: [u8; 32] = keccak256!("SubcallSucceeded");
+pub const LOG_SUBCALL_FAILED: [u8; 32] = keccak256!("SubcallFailed");
+
+pub fn log_subcall_succeeded(address: impl Into<H160>) -> Log {
+	log0(address, LOG_SUBCALL_SUCCEEDED)
+}
+
+pub fn log_subcall_failed(address: impl Into<H160>) -> Log {
+	log0(address, LOG_SUBCALL_FAILED)
+}
+
+/// Reverts if fees and gas_limit are not sufficient to make subcall and cleanup
+fn ensure_can_provide_randomness<Runtime>(
+	code_address: H160,
+	gas_limit: u64,
+	request_fee: Option<BalanceOf<Runtime>>,
+	clean_up_cost: Option<u64>,
+) -> EvmResult<()>
+where
+	Runtime: pallet_randomness::Config + pallet_evm::Config + pallet_base_fee::Config,
+	BalanceOf<Runtime>: Into<U256>,
+{
+	if let Some(fee) = request_fee {
+		// assert fee / base_fee > gasLimit
+		let fee_as_u256: U256 = fee.into();
+		// TODO: should this be as_u64 which panics?
+		let fees_available: u64 = fee_as_u256
+			.checked_div(pallet_base_fee::Pallet::<Runtime>::base_fee_per_gas())
+			.unwrap_or_default()
+			.low_u64();
+		if gas_limit >= fees_available {
+			return Err(revert(
+				"Gas limit at current price must be less than fees allotted",
+			));
+		}
+	}
+	let log_cost = log_subcall_failed(code_address)
+		.compute_cost()
+		.map_err(|_| revert("failed to compute log cost"))?;
+	// Cost of the call itself that the batch precompile must pay.
+	let call_cost = call_cost(U256::zero(), <Runtime as pallet_evm::Config>::config());
+	// assert gasLimit > overhead cost
+	let overhead = if let Some(estimated_clean_up_cost) = clean_up_cost {
+		call_cost + log_cost + estimated_clean_up_cost
+	} else {
+		call_cost + log_cost
+	};
+	if gas_limit <= overhead {
+		return Err(revert("Gas limit must exceed overhead call cost"));
+	}
+	Ok(())
+}
+
+/// Subcall to provide randomness
+/// caller must call `ensure_can_provide_randomness` before calling this function
```

Comment:
I hit this error last time I tried this:
```rust
error[E0632]: cannot provide explicit generic arguments when `impl Trait` is used in argument position
   --> precompiles/randomness/src/lib.rs:373:24
    |
373 |         provide_randomness::<Runtime>(
    |                              ^^^^^^^ explicit generic argument not allowed
```
So that is why I separated it out. I just respent some time trying to get it to work and failed, but not sure it's that important. It's only called once anyway.

We need the generic for 

```rust
   Compiling pallet-evm-precompile-randomness v0.1.0 (/Users/4meta5/ong/ps/mb/precompiles/randomness)
error[E0412]: cannot find type `Runtime` in this scope
  --> precompiles/randomness/src/lib.rs:76:59
   |
67 | fn ensure_can_provide_randomness(
   |                                 - help: you might be missing a type parameter: `<Runtime>`
...
76 |         gas_limit_as_u256.checked_mul(pallet_base_fee::Pallet::<Runtime>::base_fee_per_gas())
   |                                                                 ^^^^^^^ not found in this scope

error[E0412]: cannot find type `Runtime` in this scope
  --> precompiles/randomness/src/lib.rs:90:43
   |
90 |     let call_cost = call_cost(U256::zero(), <Runtime as pallet_evm::Config>::config());
   |                                              ^^^^^^^ not found in this scope
```